janusgraph-core/src/main/java/org/janusgraph/diskstorage/configuration/builder/ReadConfigurationBuilder.java

```diff
@@ -0,0 +1,303 @@
+                }
+
+                checkJanusGraphVersion(globalWrite, localBasicConfiguration, keyColumnValueStoreConfiguration, overwrite);
+
+                checkOptionsWithDiscrepancies(globalWrite, localBasicConfiguration, overwrite);
+            }
+
+            return keyColumnValueStoreConfiguration.asReadConfiguration();
+        }
+    }
+
+    private void setupUpgradeConfiguration(String graphName, ModifiableConfiguration globalWrite){
+
+        // If the graph doesn't have a storage version set it and update version
+        if (!globalWrite.has(INITIAL_STORAGE_VERSION)) {
+            globalWrite.set(INITIAL_JANUSGRAPH_VERSION, JanusGraphConstants.VERSION);
+            globalWrite.set(TITAN_COMPATIBLE_VERSIONS, JanusGraphConstants.VERSION);
+            globalWrite.set(INITIAL_STORAGE_VERSION, JanusGraphConstants.STORAGE_VERSION);
+            globalWrite.set(ALLOW_UPGRADE, false);
+            LOG.info("graph.storage-version has been upgraded from 1 to {} and graph.janusgraph-version has been upgraded from {} to {} on graph {}",
+                JanusGraphConstants.STORAGE_VERSION, globalWrite.get(INITIAL_JANUSGRAPH_VERSION), JanusGraphConstants.VERSION, graphName);
+            // If the graph has a storage version, but it's lower than the client or server opening the graph upgrade the version and storage version
+        } else if (Integer.parseInt(globalWrite.get(INITIAL_STORAGE_VERSION)) < Integer.parseInt(JanusGraphConstants.STORAGE_VERSION)) {
+            globalWrite.set(INITIAL_JANUSGRAPH_VERSION, JanusGraphConstants.VERSION);
+            globalWrite.set(TITAN_COMPATIBLE_VERSIONS, JanusGraphConstants.VERSION);
+            globalWrite.set(INITIAL_STORAGE_VERSION, JanusGraphConstants.STORAGE_VERSION);
+            globalWrite.set(ALLOW_UPGRADE, false);
+            LOG.info("graph.storage-version has been upgraded from {} to {} and graph.janusgraph-version has been upgraded from {} to {} on graph {}",
+                globalWrite.get(INITIAL_STORAGE_VERSION), JanusGraphConstants.STORAGE_VERSION, globalWrite.get(INITIAL_JANUSGRAPH_VERSION), JanusGraphConstants.VERSION, graphName);
+            // If the storage version of the client or server opening the graph is lower than the graph's storage version throw an exception
+        } else if (Integer.parseInt(globalWrite.get(INITIAL_STORAGE_VERSION)) > Integer.parseInt(JanusGraphConstants.STORAGE_VERSION)) {
+            throw new JanusGraphException(String.format(BACKLEVEL_STORAGE_VERSION_EXCEPTION, globalWrite.get(INITIAL_STORAGE_VERSION), JanusGraphConstants.STORAGE_VERSION, graphName));
+        } else {
+            LOG.warn("Warning graph.allow-upgrade is currently set to true on graph {}. Please set graph.allow-upgrade to false in your properties file.", graphName);
+        }
+    }
+
+    private org.apache.commons.configuration.Configuration toConfiguration(ReadConfiguration localConfig) {
```

Comment:
I will remove this method and will use the suggestion you provided, but `ReadConfigurationConverter` is still used in `GraphDatabaseConfiguration` in `getConfigurationAtOpen()` method. Of course we can move conversion logic back into `GraphDatabaseConfiguration` but I think that then it will violate the single responsibility principle. I think it still violates this principle, but less then it was. So, I would leave `ReadConfigurationConverter` if you don't mind.

Reply:
I'm fine with leaving the converter. At the end of the day `GraphDataBaseConfiguration` is still a huge class and there's nothing wrong with trying to abstract some of the logic out and I can't currently think of a better place for this function.